lib/beaker/hypervisor/aws_sdk.rb

```diff
@@ -220,125 +221,171 @@ def kill_zombie_volumes
           }
-        end
-
-        security_group = ensure_group(vpc || region, Beaker::EC2Helper.amiports(host))
-
-        # Launch the node, filling in the blanks from previous work.
-        @logger.notify("aws-sdk: Launch instance")
-        config = {
-          :count => 1,
-          :image_id => image_id,
-          :monitoring_enabled => true,
-          :key_pair => ensure_key_pair(region),
-          :security_groups => [security_group],
-          :instance_type => amisize,
-          :disable_api_termination => false,
-          :instance_initiated_shutdown_behavior => "terminate",
-          :block_device_mappings => block_device_mappings,
-          :subnet => subnet_id,
         }
-        instance = region.instances.create(config)
+      end
 
-        # Persist the instance object for this host, so later it can be
-        # manipulated by 'cleanup' for example.
-        host['instance'] = instance
+      security_group = ensure_group(vpc || region, Beaker::EC2Helper.amiports(host))
 
-        @logger.notify("aws-sdk: Launched #{host.name} (#{amitype}:#{amisize}) using snapshot/image_type #{image_type}")
-      end
+      msg = "aws-sdk: launching %p on %p using %p/%p%s" %
+            [host.name, amitype, amisize, image_type,
+             subnet_id ? ("in %p" % subnet_id) : '']
+      @logger.notify(msg)
+      config = {
+        :count => 1,
+        :image_id => image_id,
+        :monitoring_enabled => true,
+        :key_pair => ensure_key_pair(region),
+        :security_groups => [security_group],
+        :instance_type => amisize,
+        :disable_api_termination => false,
+        :instance_initiated_shutdown_behavior => "terminate",
+        :block_device_mappings => block_device_mappings,
+        :subnet => subnet_id,
+      }
+      region.instances.create(config)
+    end
 
+    # Create EC2 instances for all hosts, tag them, and wait until
+    # they're running.
+    #
+    # @return [void]
+    # @api private
+    def launch_all_nodes
+      @logger.notify("aws-sdk: launch all hosts in configuration")
+      ami_spec = YAML.load_file(@options[:ec2_yaml])["AMI"]
+      global_subnet_id = @options['subnet_id']
+      global_subnets = @options['subnet_ids']
+      if global_subnet_id && global_subnets
+        raise RuntimeError, 'Config specifies both subnet_id and subnet_ids'
+      end
+      any_subnet_hosts, other_hosts = @hosts.partition do |host|
+        global_subnets && ! (global_subnet_id || host['subnet_id'])
+      end
+      instances = [] # Each element is {:instance => i, :host => h}
+      begin
+        @logger.notify("aws-sdk: launch instances not particular about subnet")
+        # Shuffle the subnets so we don't always hit the same one
+        # first, and cycle though the subnets independently of the
+        # host, so we stick with one that's working.  Try each subnet
+        # once per-host.
+        subnet_i = 0
+        if global_subnets
```

Comment:
@anodelman ...and indeed, now I recall why I had that there -- it's to guard the subnets.shuffle (if nothing else) since subnets will be nil when not specified.  But I can change it to something like "shuffnets = !any_subnet_hosts.empty? && subnets.shuffle" or similar if you prefer.